src/components/product-progress/product-progress.repository.ts

```diff
@@ -1,7 +1,189 @@
 import { Injectable } from '@nestjs/common';
-import { DatabaseService } from '../../core';
+import { node, Query, relation } from 'cypher-query-builder';
+import { DateTime } from 'luxon';
+import { generateId, ID, NotFoundException, UnsecuredDto } from '../../common';
+import { DatabaseService, property } from '../../core';
+import { matchProps } from '../../core/database/query';
+import { ProductProgressInput, StepProgress } from './dto';
 
 @Injectable()
 export class ProductProgressRepository {
   constructor(private readonly db: DatabaseService) {}
+
+  async removeOldProgress(productId: ID, reportId: ID) {
+    await this.db
+      .query()
+      .match([
+        node('product', 'Product', { id: productId }),
+        relation('out', 'productProgressRel', 'productProgress', {
+          active: true,
+        }),
+        node('node', 'ProductProgress'),
+        relation('in', 'reportProgressRel', 'reportProgress', { active: true }),
+        node('report', 'PeriodicReport', { id: reportId }),
+      ])
+      .setValues({
+        'productProgressRel.active': false,
+        'reportProgressRel.active': false,
+      })
+      .return('node.id as id')
+      .asResult<{ id: ID }>()
+      .run();
+  }
+
+  async create(input: ProductProgressInput) {
+    const id = await generateId();
+    const createdAt = DateTime.local();
+    const query = this.db
+      .query()
+      .match([node('product', 'Product', { id: input.productId })])
+      .match([node('report', 'PeriodicReport', { id: input.reportId })])
+      .create([
+        [
+          node('node', ['ProductProgress', 'BaseNode'], {
+            id,
```

Comment:
If we are actually creating an ID we should add it the DTO. But I left it off because `ProductProgress` isn't really created. It's just an artificial grouping of each product's step's progress for the reporting period.